### PR TITLE
Support MapRoulette Templating and Mustache Tags

### LINKED ISSUE #1419: [MapRoulette Integration] Curly braces in instructions aren't replaced
### Description

I'm not sure this is something you can even fix, but I noticed that the {{curly braces}} parameters in the instructions don't get replaced with the correct values. The way that Maproulette allows people to write instructions it will replace {{idendifier}} with the value of identifier in the task. See Screen shots.

For more detail from MapRoulette: https://learn.maproulette.org/en-us/documentation/mustache-tag-replacement/#content

### Screenshots

Here is what I see on Rapid
Screenshot 2024-05-09 at 3 56 02 PM
And here is what That task looks like on MapRoulette
Screenshot 2024-05-09 at 3 57 35 PM
Here is where the "osmIdentifier" comes from
When you do a `curl -X 'GET' 'https://maproulette.org/api/v2/task/227267509' -H 'accept: application/json'`
you will see "properties". These can be referenced in the instructions by using {{}} and they are supposed to be replaced.
For this specific task the properties look like this:
```
        "properties": {
          "oneway": "yes",
          "highway": "cycleway",
          "crossing": "unmarked",
          "cycleway": "crossing",
          "priority": 0,
          "identifier": "w1204730293@1",
          "instructions": "None",
          "osmIdentifier": "way/1204730293",
          "crossing:markings": "no"
```

### Version

2.3.0

### What browser are you seeing the problem on? What version are you running?

Chrome v124.0

### The OS you're using

mac

### Steps to reproduce

Go to https://rapideditor.org/edit#map=21.97/37.77652/-122.40743&background=Bing&datasets=fbRoads,msBuildings&disable_features=boundaries and enable the MapRoulette Tasks.
Click on the one in the middle of your screen and scroll down to the instructions and look at the first sentence.

### The browser URL at the time you encountered the bug

https://rapideditor.org/edit#map=21.97/37.77652/-122.40745&background=Bing&datasets=fbRoads,msBuildings&disable_features=boundaries

### The auto-detected useragent string for your browser (leave blank if you're manually filling this form out)

Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36

### LINKED ISSUE #1423: [MapRoulette Integration] Support MapRoulette templating in challenge instructions
### Description

MapRoulette supports basic form fields and short codes in challenge instructions. E.g.:

`[checkbox "Your checkbox label" name="myCheckbox"]`

`[select "Your select label" name="myDropdown" values="foo,bar,baz"]`

`[copyable "some useful text"]`

This markup is documented here: https://learn.maproulette.org/en-us/documentation/challenge-instructions-templating/

Currently in Rapid, this markup is not parsed and is displayed as raw text in the instructions.

This issue is related to #1419 - Curly braces in instructions aren't replaced

### Screenshots

Screenshot 2024-05-13 at 6 33 57 AM

### Version

2.3.1

### What browser are you seeing the problem on? What version are you running?

Firefox 125.0.3

### The OS you're using

MacOS 14.4.1

### Steps to reproduce

1. Open Rapid and enable MapRoulette tasks in the Map Data using Challenge ID 43887
2. Zoom/Pan to a MapRoulette task in Mono County and select the task
3. Note that the MapRoulette markup in the instructions is rendered as plain text

### The browser URL at the time you encountered the bug

https://rapideditor.org/rapid#map=15.85/38.0448/-119.2625&background=Bing&datasets=fbRoads,msBuildings&disable_features=boundaries

### The auto-detected useragent string for your browser (leave blank if you're manually filling this form out)

_No response_

### PR DESCRIPTION
**Pull request by tannerwuster**
Adds support for [Templating in Challenge Instructions](https://learn.maproulette.org/en-us/documentation/challenge-instructions-templating/) and [Mustache Tag Replacement](https://learn.maproulette.org/en-us/documentation/mustache-tag-replacement/#content) 🥸

Screenshot 2024-06-27 at 16 36 50

- Fixes #1419 
- Fixes #1423

### CODE REVIEW

**Comment by tannerwuster** [comment]
Here is the templating in action! Screenshot 2024-07-03 at 16 35 38
Here is the {{osmidenifier}} mustache tag being replaced with the way id. 
Screenshot 2024-07-03 at 16 26 41

**Comment by tannerwuster** [comment]
Now when hovering over osm identifiers it ill highlight what the challenge is talking about. Clicking on it will take you to edit mode 
Screenshot 2024-07-09 at 13 57 40

File: modules/ui/maproulette_details.js [hunk 1]
```diff
@@ -1,79 +1,131 @@
...
+          .append('p')
+          .text(`${task.parentId} / ${task.id}`)
+          .selectAll('a')
+          .attr('rel', 'noopener')
+          .attr('target', '_blank');
+      }
+
+      const description = parseShortCodes(replaceMustacheTags(task.description, task));
+      const instruction = parseShortCodes(replaceMustacheTags(task.instruction, task));
+      if (task.description && !task.description.includes('Lorem')) {
```
Comment (Bonkles) [inline modules/ui/maproulette_details.js:87]: Not sure about the 'lorem' check here- is this a default string for a new, pristine task?

File: modules/ui/maproulette_details.js [hunk 2]
```diff
@@ -1,79 +1,131 @@
...
+import { utilHighlightEntities } from '../util/index.js';
+
 export function uiMapRouletteDetails(context) {
   const l10n = context.systems.l10n;
   const maproulette = context.services.maproulette;
 
   let _qaItem;
 
 
+  function parseShortCodes(text) {
```
Comment (Bonkles) [inline modules/ui/maproulette_details.js:12]: the method name could be made more meaningful- we're not just parsing the codes in this method, we're actively generating content to replace them/ transforming the text. 

A comment describing the expectations of the incoming content text (heck, even with a comment showing a full example like what @bhousel often does) would also be helpful.

File: modules/ui/maproulette_details.js [hunk 3]
```diff
@@ -1,79 +1,131 @@
...
+    let transformedText = segments[0];
+    segments.slice(1).forEach(segment => {
+      const endIndex = segment.indexOf('&quot;');
+      const dropdownName = segment.substring(0, endIndex);
+      const valuesStart = segment.indexOf('values=&quot;') + 'values=&quot;'.length;
+      const valuesEnd = segment.indexOf('&quot;', valuesStart);
+      const options = segment.substring(valuesStart, valuesEnd).split(',');
+      const dropdownHtml = `<select name="${dropdownName}"><option value=""></option>${options.map(option => `<option value="${option.trim()}">${option.trim()}</option>`).join('')}</select>`;
+      const remainder = segment.substring(valuesEnd + '&quot;'.length).trim().replace(/^\]/, ''); // Remove the first closing bracket if it exists
+      transformedText += dropdownHtml + remainder;
```
Comment (Bonkles) [inline modules/ui/maproulette_details.js:23]: This method is a bit hard to read/parse through to understand without knowing the content flowing in. Could it be made a bit more modular? perhaps a 'generateMRDropdown' method to cleanly call out inputs/outputs that are used to generate the 'dropdownHtml' on line 21.

File: modules/ui/maproulette_details.js [hunk 4]
```diff
@@ -1,79 +1,131 @@
...
+    });
+    return transformedText;
+  }
+
+
+  function replaceMustacheTags(text, task) {
+    const tagRegex = /\{\{([\w:]+)\}\}/g;
+    return text.replace(tagRegex, (match, propertyName) => {
+      if (propertyName === 'osmIdentifier' && task.title) {
+        const osmId = task.title.split('@')[0]; // Extract the full ID including the prefix
```
Comment (Bonkles) [inline modules/ui/maproulette_details.js:33]: comments like this *really* demystify the code for brand-newbies like me- more of this!

**Review by Bonkles — CHANGES_REQUESTED** [review]
A couple of small comments to aid future maintainers! like, oh, say.... *me*. :)

File: modules/ui/maproulette_details.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tannerwuster) [reply modules/ui/maproulette_details.js:87]: So I added this back when we weren’t filtering out the hidden challenges which has a big issue of showing lorem in each challenge. I can remove it now!

File: modules/ui/maproulette_details.js [hunk 5]
```diff
@@ -1,79 +1,185 @@
...
 
+  /**
+   * Generates HTML for a dropdown menu with the specified name and options.
+   *
+   * @param {string} dropdownName The name attribute for the dropdown.
+   * @param {Array<string>} options An array of options to be included in the dropdown.
+   * @returns {string} HTML string representing a dropdown menu.
+   */
+  function generateDropdownHtml(dropdownName, options) {
+    return `<select name="${dropdownName}"><option value=""></option>${options.map(option => `<option value="${option.trim()}">${option.trim()}</option>`).join('')}</select>`;
```
Comment (bhousel) [inline modules/ui/maproulette_details.js:20]: Seems ok, but personally I'd break this up - that's a lot for a one-liner!

File: modules/ui/maproulette_details.js (hunk 3 above)
Comment (bhousel) [reply modules/ui/maproulette_details.js:46]: Oh haha I guess this is the same thing I commented on, just moved to a new function.

**Pull request merged**